api/desecapi/replication.py

```diff
@@ -0,0 +1,148 @@
+
+        if rcode != 0:
+            raise GitRepositoryException(f'{cmd} returned nonzero error code', cmd=cmd, rcode=rcode, stdout=stdout, stderr=stderr)
+
+        print('\n'.join(['<<< ' + s for s in stdout.split('\n')]))
+        return stdout
+
+    def commit_all(self, msg=None):
+        self._git('add', '.')
+        if self._git('diff', '--numstat', '--staged'):
+            self._git('commit', '-m', msg or 'update')
+
+    def init(self, name='deSEC', email='git@desec.internal'):
+        self._git('init')
+        self._git('config', 'user.name', name)
+        self._git('config', 'user.email', email)
+        # TODO on commit, git tries to access $HOME/.config/git
+        #  but due to celery privilege drop, $HOME is still /root
+        #  and this gives a 'permission denied' warning in the log
+
+    def get_head(self):
+        return self.get_commit('HEAD')
+
+    def get_commit(self, rev):
+        try:
+            commit_msg = self._git('show', rev, '--format=%s', '-s')[:-1]
+            commit_hash = self._git('show', rev, '--format=%H', '-s')[:-1]
+            return commit_msg, commit_hash
+        except GitRepositoryException as e:
+            return None, None
+
+    def remove_history(self, before: timedelta):
+        rev = self._git('log', f'--before={(datetime.now() - before).isoformat()}', '-1', '--format=%H')
+        with open(os.path.join(self.path, '.git', 'shallow'), 'w') as f:
+            f.writelines([rev])
+        self._git('reflog', 'expire', '--expire=now', '--all')
+        self._git('gc', '--prune=now')
+
+
+class ZoneRepository(Repository):
+    AXFR_SOURCE = '172.16.1.11'
+
+    def __init__(self, path):
+        super().__init__(path)
+        if not os.path.exists(os.path.join(self.path, '.git')):
+            self.init()
+            self.commit_all(msg='Inception or Recovery')
+            self.full_refresh()
+            self.commit_all(msg='Full Refresh')
+
+    def full_refresh(self):
+        print('WARNING: Full refresh of replication zone repository!')
+        for name in models.Domain.objects.values_list('name', flat=True):
+            self.refresh(name)
+
+    def refresh(self, name):
+        if '/' in name or '\x00' in name:
+            raise UnsupportedZoneNameException
+
+        # obtain AXFR
+        timeout = 60  # if AXFR take longer, the timeout must be increased (see also settings.py)
+        t = datetime.now()
+        xfr = list(dns.query.xfr(self.AXFR_SOURCE, name, timeout=timeout))
+        print(f'AXFR took {(datetime.now() - t).total_seconds():.3f}s')
+        t = datetime.now()
+        z = dns.zone.from_xfr(xfr)
+        print(f'Parsing took {(datetime.now() - t).total_seconds():.3f}s')
+
+        # TODO sort AXFR? (but take care with SOA)
+        # TODO do not parse AXFR to save time
+
+        # write zone file
+        filename = os.path.join(self.path, name + '.zone')
+        with open(filename + '~', 'w') as f:
```

Comment:
This method runs one-at-a-time; if it breaks while writing the ~-file it can (and will) be safely ignored. Subsequent code will overwrite the ~-file. I guess a file lock would achieve the same thing while being more complicated